ambry-protocol/src/main/java/com.github.ambry.protocol/CompositeSend.java

```diff
@@ -24,33 +26,48 @@
  */
 public class CompositeSend implements Send {
 
-  private final List<Send> compositSendList;
+  private final List<Send> compositeSendList;
   private long totalSizeToWrite;
   private int currentIndexInProgress;
 
-  public CompositeSend(List<Send> compositSendList) {
-    this.compositSendList = compositSendList;
+  public CompositeSend(List<Send> compositeSendList) {
+    this.compositeSendList = compositeSendList;
     this.currentIndexInProgress = 0;
-    for (Send messageFormatSend : compositSendList) {
+    for (Send messageFormatSend : compositeSendList) {
       totalSizeToWrite += messageFormatSend.sizeInBytes();
     }
   }
 
   @Override
   public long writeTo(WritableByteChannel channel) throws IOException {
     long written = 0;
-    if (currentIndexInProgress < compositSendList.size()) {
-      written = compositSendList.get(currentIndexInProgress).writeTo(channel);
-      if (compositSendList.get(currentIndexInProgress).isSendComplete()) {
+    if (currentIndexInProgress < compositeSendList.size()) {
+      written = compositeSendList.get(currentIndexInProgress).writeTo(channel);
+      if (compositeSendList.get(currentIndexInProgress).isSendComplete()) {
         currentIndexInProgress++;
       }
     }
     return written;
   }
 
+  @Override
+  public void writeTo(AsyncWritableChannel channel, Callback callback) throws IOException {
+    int lastIndex = compositeSendList.size() - 1;
+    int i = 0;
+    for (Send send : compositeSendList) {
+      if (i == lastIndex) {
+        // only the last one pass in callback
+        send.writeTo(channel, callback);
+      } else {
+        send.writeTo(channel, null);
+      }
+      i++;
+    }
```

Comment:
What's your suggestion on how to stop it?

Reply:
Had offline chat. Looks like there is no immediate way to stop the loop, because exception happens in callback.  But will enhance callback logic in next PR.